crates/cli/src/subcommands/subscribe.rs

```diff
@@ -0,0 +1,276 @@
+                    let row = serde::de::DeserializeSeed::deserialize(sats::de::serde::SeedWrapper(table_ty), op.row)?;
+                    let row = table_ty.with_value(&row);
+                    let row = serde_json::to_value(sats::ser::serde::SerializeWrapper::from_ref(&row))?;
+                    match op.op {
+                        Op::Delete => deletes.push(row),
+                        Op::Insert => inserts.push(row),
+                    }
+                }
+                Ok((upd.table_name.into(), SubscriptionTable { deletes, inserts }))
+            })
+            .collect()
+    }
+}
+
+#[derive(serde::Serialize)]
+struct SubscriptionTable {
+    deletes: Vec<serde_json::Value>,
+    inserts: Vec<serde_json::Value>,
+}
+
+#[derive(serde::Deserialize)]
+struct Event {
+    message: String,
+}
+
+#[derive(serde::Deserialize)]
+enum ServerMessage<'a> {
+    SubscriptionUpdate(#[serde(borrow)] SubscriptionUpdateJson<'a>),
+    TransactionUpdate {
+        subscription_update: SubscriptionUpdateJson<'a>,
+        event: Event,
+    },
+    #[serde(other, deserialize_with = "serde_with::rust::deserialize_ignore_any")]
+    Unknown,
+}
+
+pub async fn exec(config: Config, args: &ArgMatches) -> Result<(), anyhow::Error> {
+    let queries = args.get_many::<String>("query").unwrap();
+    let num = args.get_one::<u32>("num").copied();
+    let timeout = args.get_one::<u32>("timeout").copied();
+    let print_initial_update = args.get_flag("print_initial_update");
+
+    let conn = parse_req(config, args).await?;
+    let api = ClientApi::new(conn);
+    let module_def = api.module_def().await?;
+
+    let mut uri = http::Uri::try_from(api.con.db_uri("subscribe")).unwrap().into_parts();
+    uri.scheme = uri.scheme.map(|s| {
+        if s == Scheme::HTTP {
+            "ws".parse().unwrap()
+        } else if s == Scheme::HTTPS {
+            "wss".parse().unwrap()
+        } else {
+            s
+        }
+    });
+
+    let mut req = http::Uri::from_parts(uri).unwrap().into_client_request()?;
+    req.headers_mut().insert(header::SEC_WEBSOCKET_PROTOCOL, TEXT_PROTOCOL);
+    if let Some(auth_header) = &api.con.auth_header {
+        req.headers_mut()
+            .insert(header::AUTHORIZATION, auth_header.try_into().unwrap());
+    }
+    let (mut ws, _) = tokio_tungstenite::connect_async(req).await?;
+
+    let task = async {
+        let msg = ClientMessage::Subscribe {
+            query_strings: queries.cloned().collect(),
+        };
+        ws.send(serde_json::to_string(&msg).unwrap().into()).await?;
+
+        let mut stdout = tokio::io::stdout();
+        while let Some(msg) = ws.try_next().await? {
+            let Some(msg) = parse_msg_json(&msg) else { continue };
+            match msg {
+                ServerMessage::SubscriptionUpdate(sub) => {
+                    if print_initial_update {
+                        let output = serde_json::to_string(&sub.reformat(&module_def)?)? + "\n";
```

Comment:
This seems broadly sensible. I'm not sure there's utility to the `bsatn` format, and I would name the current/default format `table`, not `plain`, but I think you're right that we should offer at least both of JSON and tabular output. I would also like to be careful that we use the same code for formatting in both subcommands so that we get the same format.

Reply:
Note: this is slightly less trivial than expected. Because:

- the transaction updates contain deleted and inserted rows, which arguably is information we want to preserve
- the subscription update supposedly only contains inserts -- but why would we format the subscription update differently from the transaction updates?
- both of the updates refer to several tables, but it's not obvious how to display that (i.e. add some kind of "title" to a table)

What could work is to render _one_ table per update, where the rows get two additional columns "table_name STRING" and "op ENUM('I', 'D')".

Reply:
I'll submit this as a follow-up.